include/nbl/asset/CCPUMeshPacker.h

```diff
@@ -71,38 +71,142 @@ class CCPUMeshPacker final : public IMeshPacker<ICPUMeshBuffer, MDIStructType>
+CCPUMeshPacker<MDIStructType>::CCPUMeshPacker(const SVertexInputParams& preDefinedLayout, const AllocationParams& allocParams, uint16_t minTriangleCountPerMDIData, uint16_t maxTriangleCountPerMDIData)
+	:IMeshPacker<ICPUMeshBuffer, MDIStructType>(minTriangleCountPerMDIData, maxTriangleCountPerMDIData),
+	 m_allocParams(allocParams),
+	 m_perInsVtxBuffAlctrResSpc(nullptr)
+	 
+{
+	m_outVtxInputParams.enabledAttribFlags = preDefinedLayout.enabledAttribFlags;
+	m_outVtxInputParams.enabledBindingFlags = preDefinedLayout.enabledAttribFlags;
+	memcpy(m_outVtxInputParams.attributes, preDefinedLayout.attributes, sizeof(m_outVtxInputParams.attributes));
+
+	//1 attrib enabled == 1 binding
+	for (uint16_t attrBit = 0x0001, location = 0; location < SVertexInputParams::MAX_ATTR_BUF_BINDING_COUNT; attrBit <<= 1, location++)
+	{
+		if (m_outVtxInputParams.enabledAttribFlags & attrBit)
+		{
+			m_outVtxInputParams.attributes[location].binding = location;
+			m_outVtxInputParams.attributes[location].relativeOffset = 0u;
+			m_outVtxInputParams.bindings[location].stride = getTexelOrBlockBytesize(static_cast<E_FORMAT>(m_outVtxInputParams.attributes[location].format));
+			m_outVtxInputParams.bindings[location].inputRate = preDefinedLayout.bindings[preDefinedLayout.attributes[location].binding].inputRate;
+		}
+	}
+
+	m_vtxSize = calcVertexSize(preDefinedLayout, E_VERTEX_INPUT_RATE::EVIR_PER_VERTEX);
+	//TODO: allow for mesh buffers with only per instance parameters
+	assert(m_vtxSize);
+
+	m_perInstVtxSize = calcVertexSize(preDefinedLayout, E_VERTEX_INPUT_RATE::EVIR_PER_INSTANCE);
+	if (m_perInstVtxSize)
+	{
+		m_perInsVtxBuffAlctrResSpc = _NBL_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(alignof(std::max_align_t), allocParams.perInstanceVertexBuffSupportedCnt, allocParams.perInstanceVertexBufferMinAllocSize), _NBL_SIMD_ALIGNMENT);
+		_NBL_DEBUG_BREAK_IF(m_perInsVtxBuffAlctrResSpc == nullptr);
+		assert(m_perInsVtxBuffAlctrResSpc != nullptr);
+		m_perInsVtxBuffAlctr = core::GeneralpurposeAddressAllocator<uint32_t>(m_perInsVtxBuffAlctrResSpc, 0u, 0u, alignof(std::max_align_t), allocParams.perInstanceVertexBuffSupportedCnt, allocParams.perInstanceVertexBufferMinAllocSize);
+	}
+
+	initializeCommonAllocators(
+		{
+			allocParams.indexBuffSupportedCnt,
+			m_vtxSize ? allocParams.vertexBuffSupportedCnt : 0ull,
+			allocParams.MDIDataBuffSupportedCnt,
+			allocParams.indexBufferMinAllocSize,
+			allocParams.vertexBufferMinAllocSize,
+			allocParams.MDIDataBuffMinAllocSize
+		}
+	);
+}
+
 template <typename MDIStructType>
 //`Iterator` may be only an Iterator or pointer to pointer
 template <typename Iterator>
-MeshPackerBase::ReservedAllocationMeshBuffers CCPUMeshPacker<MDIStructType>::alloc(const Iterator begin, const Iterator end)
+typename CCPUMeshPacker<MDIStructType>::ReservedAllocationMeshBuffers CCPUMeshPacker<MDIStructType>::alloc(const Iterator begin, const Iterator end)
 {
 	/*
 	Requirements for input mesh buffers:
```

Comment:
btw in example 22 (raytraced AO) I've made this function `homogenizePrimitiveTypeAndIndices`

Don't use it... but have a look at it.

Reply:
TODO